pennylane/workflow/qnode.py

```diff
@@ -110,15 +110,33 @@ def _to_qfunc_output_type(
     if has_partitioned_shots:
         return tuple(_to_qfunc_output_type(r, qfunc_output, False) for r in results)
 
+    qfunc_output_leaves, qfunc_output_structure = qml.pytrees.flatten(
+        qfunc_output, is_leaf=lambda obj: isinstance(obj, (qml.measurements.MeasurementProcess))
+    )
+    num_of_measurements = len(qfunc_output_leaves)
+
+    results_leaves, results_structure = qml.pytrees.flatten(results)
+    num_of_results = len(results_leaves)
+
     # Special case of single Measurement in a list
     if isinstance(qfunc_output, list) and len(qfunc_output) == 1:
         results = [results]
 
-    # If the return type is not tuple (list or ndarray) (Autograd and TF backprop removed)
-    if isinstance(qfunc_output, (tuple, qml.measurements.MeasurementProcess)):
+    if isinstance(qfunc_output, qml.measurements.MeasurementProcess):
         return results
 
-    return type(qfunc_output)(results)
+    if num_of_measurements != num_of_results or (
+        qfunc_output_structure.is_leaf != results_structure.is_leaf

```

Comment:
Also, I'm a little unsure about making sure that both are leaves.  For example, we should be able to do:


```
@qml.qnode(qml.device('default.qubit'))
def circuit():
    return {"m": qml.expval(qml.Z(0))}
circuit()
```
```
File [~/Prog/pennylane/pennylane/workflow/qnode.py:134](http://localhost:8888/~/Prog/pennylane/pennylane/workflow/qnode.py#line=133), in _to_qfunc_output_type(results, qfunc_output, has_partitioned_shots)
    131     if isinstance(qfunc_output, (tuple)):
    132         # FIXME: Work around for Autograd and TF backprop
    133         return results
--> 134     return type(qfunc_output)(results)
    136 if isinstance(qfunc_output, qml.typing.TensorLike):
    137     return type(qfunc_output)(results)

TypeError: 'numpy.float64' object is not iterable
```